### PR TITLE
Feature: dynamic shared mem moe_align_block_size_kernel

### PR DESCRIPTION
**Pull request by akhoroshev**
I encountered compilation errors related to insufficient shared memory size when I tried to increase NUM_MAX_EXPERTS to 128.

@zwd003

### CODE REVIEW

**Comment by WoosukKwon** [comment]
@pcmoritz Can you take a look? The PR looks good to me.

**Review by pcmoritz — APPROVED** [review]
Thanks, the PR looks good to me, feel free to merge @WoosukKwon !

File: csrc/moe_align_block_size_kernels.cu [hunk 1]
```diff
@@ -7,10 +7,16 @@
...
 
-const static size_t NUM_MAX_EXPERTS = 64;
 #define CEILDIV(x,y) (((x) + (y) - 1) / (y))
 
 namespace vllm {
+
+namespace {
+__device__ int32_t index(int32_t total_col, int32_t row, int32_t col) {
+    return row * total_col + col;
+}
```
Comment (WoosukKwon) [inline csrc/moe_align_block_size_kernels.cu:17]: nit: 
```suggestion
namespace {
inline __device__ int32_t index(int32_t total_col, int32_t row, int32_t col) {
    return row * total_col + col;
}
```

File: csrc/moe_align_block_size_kernels.cu [hunk 2]
```diff
@@ -93,11 +103,16 @@ void moe_align_block_size(
...
     VLLM_DISPATCH_INTEGRAL_TYPES(
         topk_ids.scalar_type(), "moe_align_block_size_kernel", [&] {
-        vllm::moe_align_block_size_kernel<scalar_t><<<1, num_experts, 0, stream>>>(
-            topk_ids.data_ptr<scalar_t>(), 
+        // calc needed amount of shared mem for `tokens_cnts` and `cumsum` tensors
+        const int32_t shared_mem = ((num_experts + 1) * num_experts + (num_experts + 1)) * sizeof(int32_t);
+
+        // set dynamic shared mem
+        auto kernel = vllm::moe_align_block_size_kernel<scalar_t>;
+        TORCH_CHECK(cudaFuncSetAttribute(kernel, cudaFuncAttributeMaxDynamicSharedMemorySize, shared_mem) == cudaSuccess);
```
Comment (WoosukKwon) [inline csrc/moe_align_block_size_kernels.cu:113]: nit: Can we do something like this?
```suggestion
        TORCH_CUDA_CHECK(cudaFuncSetAttribute(kernel, cudaFuncAttributeMaxDynamicSharedMemorySize, shared_mem));
```

**Review by WoosukKwon — COMMENTED** [review]
@akhoroshev Thanks for submitting the PR! Left minor comments.

**Comment by WoosukKwon** [comment]
Also, just curious: which MoE model are you using? Is there any public model with more than 128 experts?

**Comment by akhoroshev** [comment]
> Also, just curious: which MoE model are you using? Is there any public model with more than 128 experts?

This is not a public model

**Comment by akhoroshev** [comment]
@WoosukKwon comments taken into account

**Review by WoosukKwon — APPROVED** [review]
@akhoroshev  LGTM! Thanks for the PR!

**Pull request merged**